src/clustering/administration/servers/auto_reconnect.cc

```diff
@@ -82,24 +82,48 @@ void auto_reconnector_t::try_reconnect(const server_id_t &server,
         },
         initial_call_t::YES);
 
-    wait_any_t interruptor(&reconnected, &give_up_timer, keepalive.get_drain_signal());
-
     cond_t join_failed;
+    wait_any_t interruptor(
+        &reconnected, &give_up_timer, keepalive.get_drain_signal(), &join_failed);
     exponential_backoff_t backoff(50, 15 * 1000);
     try {
-        while (!interruptor.is_pulsed() && !join_failed.is_pulsed()) {
-            coro_t::spawn_now_dangerously([&]() {
-                join_result_t result =
+        // These can be safely passed into the coroutine below.
+        // They will be reset to `nullptr` by the assignment_sentry_ts when this function
+        // call ends.
+        std::shared_ptr<cond_t *> join_failed_out(new cond_t *(nullptr));
+        std::shared_ptr<peer_address_t *> last_known_address_out(new peer_address_t *(nullptr));
+        assignment_sentry_t<cond_t *> join_failed_out_assignment(join_failed_out.get(), &join_failed);
+        assignment_sentry_t<peer_address_t *> last_known_address_out_assignment(last_known_address_out.get(), &last_known_address);
+
+        while (!interruptor.is_pulsed()) {
+            // This coroutine can keep running even after this function has returned
+            // or the auto_reconnector_t was destructed. So we must pass everything
+            // we need by value.
+
+            coro_t::spawn_now_dangerously(
+                [this, last_known_address, server, join_failed_out, last_known_address_out]() {
+                auto &connectivity_cluster_run_local = connectivity_cluster_run;
+                auto &join_delay_secs_local = join_delay_secs;
+
+                join_results_t results =
                     connectivity_cluster_run->join_blocking(
```

Comment:
I think this should use `connectivity_cluster_run_local` too.
I'm actually not sure if we need `connectivity_cluster_run_local`. We could maybe just use `connectivity_cluster_run` here, though using the `local` one is probably not a bad idea.